File: src/github_to_eat/preflight.py
"""Preflight checks: read-only validation that runs before any import writes."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from .client import EATClient


@dataclass(frozen=True)
class PreflightResult:
    project_id: int
    project_title: str
    non_empty: bool


def _project_title(project: dict[str, Any], project_id: int) -> str:
    # EAT returns the project name in ``project_title``; keep title/name as fallbacks.
    for key in ("project_title", "title", "name"):
        value = project.get(key)
        if value:
            return str(value)
    return f"project {project_id}"


def preflight(client: EATClient, project_id: int) -> PreflightResult:
    """Confirm the API/token work and the project is reachable; flag if non-empty.

    Runs the connectivity check first so an invalid token fails fast before we
    touch the project.
    """
    client.get_meta()
    project = client.get_project(project_id)
    return PreflightResult(
        project_id=project_id,
        project_title=_project_title(project, project_id),
        non_empty=client.project_has_stories(project_id),
    )

File: src/preflight.js
/** Preflight checks: read-only validation that runs before any import writes. */

/**
 * The subset of the client the preflight needs (kept structural for tests).
 *
 * @typedef {object} PreflightClient
 * @property {() => Promise<any>} getMeta
 * @property {(projectId: number) => Promise<any>} getProject
 * @property {(projectId: number) => Promise<boolean>} projectHasStories
 */

/**
 * @typedef {object} PreflightResult
 * @property {number} projectId
 * @property {string} projectTitle
 * @property {boolean} nonEmpty
 */

/**
 * @param {any} project
 * @param {number} projectId
 * @returns {string}
 */
function projectTitle(project, projectId) {
  // EAT returns the project name in `project_title`; keep title/name as fallbacks.
  for (const key of ["project_title", "title", "name"]) {
    const value = project[key];
    if (value) return String(value);
  }
  return `project ${projectId}`;
}

/**
 * Confirm the API/token work and the project is reachable; flag if non-empty.
 *
 * Runs the connectivity check first so an invalid token fails fast before we
 * touch the project.
 *
 * @param {PreflightClient} client
 * @param {number} projectId
 * @returns {Promise<PreflightResult>}
 */
export async function preflight(client, projectId) {
  await client.getMeta();
  const project = await client.getProject(projectId);
  return {
    projectId,
    projectTitle: projectTitle(project, projectId),
    nonEmpty: await client.projectHasStories(projectId),
  };
}
